### PR TITLE
WIP: Require an automatic parameter name to be a Python identifier

### LINKED ISSUE #2483: Missing auto-generated environment variables in help screen & case-sensitivity
Here is a simple CLI that mixes auto-generated and multiple environment variables:
```python
from click import command, echo, option, pass_context

@command(context_settings={"auto_envvar_prefix": "yo", "show_default": True})
@option("--flag/--no-flag", envvar=["FlAg", "sUper"], show_envvar=True)
@pass_context
def my_cli(ctx, flag):
    echo(f"Flag value: {flag}")

if __name__ == "__main__":
    my_cli()
```

This produces the following help screen:
```shell-session
$ python ./multiple_envvars.py --help
Usage: multiple_envvars.py [OPTIONS]

Options:
  --flag / --no-flag  [env var: FlAg, sUper; default: no-flag]
  --help              Show this message and exit.
```

Now let's try a couple of different envvars:
```shell-session
$ FlAg=1 python ./multiple_envvars.py
Flag value: True

$ FlAg=0 python ./multiple_envvars.py
Flag value: False
```

```shell-session
$ sUper=1 python ./multiple_envvars.py
Flag value: True

$ sUper=0 python ./multiple_envvars.py
Flag value: False
```

```shell-session
$ yo_FLAG=1 python ./multiple_envvars.py
Flag value: False

$ YO_FLAG=1 python ./multiple_envvars.py
Flag value: True

$ YO_FLAG=0 python ./multiple_envvars.py
Flag value: False
```

The surprise here is the discovery process of the auto-generated `YO_FLAG`. I expected it to respect the case-sensitivity of others. but everything auto-generated is uppercased.

So I wonder if we should enforce upper-casing all env vars, whatever their source.

And additionally, shouldn't we add the auto-generated `YO_FLAG` to the help screen to improve discoverability?

For reference, I ran these test on macOS / zsh, and with the latest `click 8.3.1`.

### PR DESCRIPTION
**Pull request by kdeldycke**
> [!WARNING]
> This PR is **waiting for #3866** to be settled/merged into the 8.x release cycle to be rebased on it, as both shares extensive unit tests.

This PR changes the behavior of automatic naming:
- An argument must produce a valid Python identifier. `0-file`, `foo.bar` and `foo bar` are now refused.
- `expose_value=False` no longer bypasses that check, on both arguments and options.
- An argument or option requires at least one declaration that can't be empty. `click.argument(expose_value=False)` and `click.option(expose_value=False)` used to produce a `""` parameter.
- An awkward explicit option name is normalized: `click.option("--in-file", "Input_File")` names its parameter `input_file`.

What does not change:

- The transformation itself: every `-` becomes `_` and the result is lower cased.
- An exposed option still raise if not producing a Python identifier, like `click.option("--0-file")`.

How to migrate:

- Give an option an explicit name: `click.option("--0-file", "zero_file")`.
- Rename an argument and pass `metavar` to keep the old display: `click.argument("zero_file", metavar="0-FILE")`.

This PR shares the majority of the tests from #3866 which lockdown all other edge-cases and canonical naming behavior.

### Context

It is in the same vein as my previous PR at https://github.com/pallets/click/pull/3808 , and is also based on a collection of edge-cases I accumulated over the years. It also covers platform differences between Windows and Unix-like (see: https://github.com/pallets/click/issues/2483).

Note: this PR started as simple test coverage expansion but ended up implementing sanitizing fixes.

### CODE REVIEW

**Comment by davidism** [comment]
Both of these seem like bugs. I feel like the automatic names should always be normalized and pass `str.isidentifier`.

**Comment by kdeldycke** [comment]
> Both of these seem like bugs. I feel like the automatic names should always be normalized and pass `str.isidentifier`.

OK cool, that was also my feeling. We don't want special treatments of the `name` of parameters, whatever their kind (options or arguments). I cannot find any reasons with the transformation should be different between the two kinds. So in the end yes, `str.isidentifier` should be the minimal check common to the two.

**Comment by kdeldycke** [comment]
> Both of these seem like bugs. I feel like the automatic names should always be normalized and pass `str.isidentifier`.

@davidism Done. I did not pushed the normalization too much.

So there is one bothering difference between `Argument` and `Option`: the latter makes `click.option("--in-file", "Input_File")` names its parameter `Input_File`, while Argument forces it to `input_file`. Should I align both? I would tend to yes.

**Comment by kdeldycke** [comment]
I just pushed the alignment and normalization of naming a bit more and updated the differences in the body of that PR.

**Comment by kdeldycke** [comment]
OK these were my final edits. You can review this PR.

File: docs/arguments.md [hunk 1]
```diff
@@ -50,6 +50,80 @@ recognized, otherwise {data}`STRING` is used. If no default value is
...
+```{caution}
+Only the ASCII hyphen is replaced, so a separator that merely looks like one is
+refused, as is any character that renders nothing.
+
+The identifier set itself moves with the Unicode table Python ships: the
+zero-width joiner (`U+200D`) entered it in Python 3.13, so a declaration holding
+one is refused up to Python 3.12 and names a parameter from 3.13 on. Prefer a
+declaration that is already a lower-case identifier with `-` for `_`.
+```
+
```
Comment (davidism) [inline docs/arguments.md:121]: This is probably too much detail. I doubt anyone's writing these, but linking to `isidentifier` should already be enough.

File: docs/options.md [hunk 2]
```diff
@@ -87,12 +83,60 @@ characters are replaced with `_`.
...
+read the other's value. Pass an explicit name instead:
+`click.option("--0-file", "zero_file", expose_value=False)`.
+
+```{caution}
+Transformation from option name to argument name is not reversible. And is many-to-one: several option names can map to the same argument name.
+
+For example, `--foo-bar`, `--Foo-Bar` and `--FOO-BAR` all map to `foo_bar`.
+
+This is allowed so that options can deliberately form a [feature switch group](#feature-switch-group).
 ```
```
Comment (davidism) [inline docs/options.md:140]: This also seems like too much detail.

The example at the end about capitalization can be added to the table at the top if it's not there.

File: CHANGES.md [hunk 3]
```diff
@@ -2,6 +2,14 @@
 
 Unreleased
 
+- {class}`Argument` derives its name the same way {class}`Option` does, and must be
+  a valid Python identifier, else raises `TypeError`. {pr}`3827`
+- `expose_value=False` no longer excuses that check on either kind. Pass an
+  explicit name to {class}`Option`, or rename an {class}`Argument` and pass
+  `metavar` to keep its display. {pr}`3827`
+- Neither kind builds a parameter without a declaration. `click.argument()` and
```
Comment (davidism) [inline CHANGES.md:10]: Is this saying that `Parameter.name` can no longer be `None`? Do we need to update the annotation? Would be great if so.

**Comment by davidism** [comment]
This needs to be targeted at 9.0 and listed high up in the changelog, the name normalization is a breaking change. Along with the keyword check I mention below, this could be significantly disruptive. We should probably issue a deprecation warning and accept names that are currently accepted, before applying the new rules.

It might be good for the `TypeError` to give explicit instructions: `Argument {decls} tried to use 'name' but it is not a valid Python identifier. Add a valid name to the parameter declaration.`

Overall I think the docs are a little _too_ verbose about this. Giving a few examples and linking to `isidentifier` should be enough.

You bring up a good point about keywords. I think the intention of this code was clear: to only accept names that can be used in signatures rather than `kwargs`. We can also check that the name isn't a keyword with https://docs.python.org/3/library/keyword.html. I bet `from` and `in` are pretty common option names though, perhaps this is not worth it?

File: src/click/core.py [hunk 4]
```diff
@@ -2436,9 +2434,47 @@ def __repr__(self) -> str:
...
+
+        :raises TypeError: when no name was derived, or the one derived is not
+            an identifier.
+        """
+        if name is not None and name.isidentifier():
+            return name
+
+        raise TypeError(
+            _(
+                "Could not determine name for {param_type} with declarations {decls!r}"
```
Comment (davidism) [inline src/click/core.py:2474]: "Could not determine a valid Python identifier for param_type decls."

**Review by Rowlando13 — COMMENTED** [review]
I agree with @davidism. The docs are too verbose and these changes should go through normal deprecation process. Probably a best course of action is to make a pr against stable that has whatever tests you think we should have to lock in existing behavior. Then can merge stable into main. And you can make the new pr against main with the deprecations.

**Comment by kdeldycke** [comment]
Ahah agreed 100% about the verbosity of the docs. They accumulates all my changes and discovery through the process of producing this PR. It was a way for me to highlight how clunky the ID derivation is at the moment. My thinking being that if I can write the docs easily and make them flow naturally, it probably means the underlying concept and implementation is clean.

Ok for a 9.0.0 target then.

What I will do:
- keep targeting this for a clean break for 9.0.0
- propose another PR for 8.5.1/8.6.0 with proper detection of the edge-cases, deprecation warnings and migration documentation so the future break has time to diffuse in the ecosystem

File: docs/arguments.md (hunk 1 above)
Comment (kdeldycke) [reply docs/arguments.md]: > This is probably too much detail. I doubt anyone's writing these, but linking to `isidentifier` should already be enough.

Agreed. These details are materialized in the test suite anyway so no need for this level of verbosity. Removed.

File: docs/options.md (hunk 2 above)
Comment (kdeldycke) [reply docs/options.md]: I was able to simplify these explanation with examples added to the table, but more importantly, but aligning the Option and Argument processing strictly, and moving the whole explanation to a dedicated section to the common `parameters.md` docs.

File: CHANGES.md (hunk 3 above)
Comment (kdeldycke) [reply CHANGES.md]: Ah good catch! And yes, `Parameter.name` is declared `name: str` at the top level. And your comment made me realized I could simplify some annotation and branch checks. Done.

File: src/click/core.py (hunk 4 above)
Comment (kdeldycke) [reply src/click/core.py]: Fixed.

**Comment by kdeldycke** [comment]
> You bring up a good point about keywords. I think the intention of this code was clear: to only accept names that can be used in signatures rather than `kwargs`. We can also check that the name isn't a keyword with https://docs.python.org/3/library/keyword.html. I bet `from` and `in` are pretty common option names though, perhaps this is not worth it?

Problem is the [list of Python keywords](https://docs.python.org/3/reference/lexical_analysis.html#keywords) would prevent some options like `--import` or `--from`. Which is a hard constraint to impose a developer of a CLI. I prefer to not enforce checks too much here.

**Comment by kdeldycke** [comment]
OK so I think I am done with this PR and it now implements the ideal target behavior for 9.0.0. I will now create in the next few days a sibling PR for 8.5.1/8.6.0 that detects the changes and emits deprecation warnings. If I discover new stuff while producing this other PR I might bring some changes here.
